batchproducer/batchproducer.go

```diff
@@ -0,0 +1,314 @@
+// or greater than bufferSize then the Add method will block.
+// TODO: this is too many args. Maybe instead have some defaults and add some setter methods for the less critical
+// params
+func NewBatchProducer(
+	client BatchingKinesisClient,
+	streamName string,
+	bufferSize int,
+	flushInterval time.Duration,
+	batchSize int,
+	logger *log.Logger,
+) (BatchProducer, error) {
+	if batchSize < 1 || batchSize > 500 {
+		return nil, errors.New("batchSize must be between 1 and 500 inclusive")
+	}
+
+	if bufferSize < batchSize && flushInterval <= 0 {
+		return nil, errors.New("If bufferSize < batchSize && flushInterval <= 0 then the buffer will eventually fill up and Add will block forever.")
+	}
+
+	if flushInterval > 0 && flushInterval < 50*time.Millisecond {
+		return nil, errors.New("Are you crazy?")
+	}
+
+	batchProducer := batchProducer{
+		client:               client,
+		streamName:           streamName,
+		flushInterval:        flushInterval,
+		batchSize:            batchSize,
+		maxAttemptsPerRecord: 10,
+		logger:               logger,
+		statInterval:         time.Second,
+		currentStat:          new(BatchStat),
+		records:              make(chan batchRecord, bufferSize),
+		stop:                 make(chan interface{}),
+	}
+
+	return &batchProducer, nil
+}
+
+type batchProducer struct {
+	client               BatchingKinesisClient
+	streamName           string
+	flushInterval        time.Duration
+	batchSize            int
+	maxAttemptsPerRecord int
+	logger               *log.Logger
+	running              bool
+	consecutiveErrors    int
+	currentDelay         time.Duration
+	statInterval         time.Duration
+	statReceiver         BatchStatReceiver
+	currentStat          *BatchStat
+	records              chan batchRecord
+	stop                 chan interface{}
+}
+
+type batchRecord struct {
+	data         []byte
+	partitionKey string
+	sendAttempts int
+}
+
+// from/for interface BatchProducer
+func (b *batchProducer) Add(data []byte, partitionKey string) error {
+	if !b.running {
+		return errors.New("Cannot call Add when BatchProducer is not running (to prevent the buffer filling up and Add blocking indefinitely).")
+	}
+	b.records <- batchRecord{data: data, partitionKey: partitionKey}
+	return nil
+}
+
+// from/for interface BatchProducer
+func (b *batchProducer) Start() error {
+	if b.running {
+		return nil
+	}
+
+	go b.run()
+
+	for !b.running {
+		// Give the goroutine a chance to start before returning.
+		time.Sleep(1 * time.Microsecond)
+	}
+
+	return nil
+}
+
+func (b *batchProducer) run() {
+	flushTick := time.Tick(b.flushInterval)
+
+	var statTick <-chan time.Time
+	if b.statReceiver != nil {
+		statTick = time.Tick(b.statInterval)
+	}
+
+	b.running = true
+	defer func() { b.running = false }()
+
+	for {
+		select {
+		case <-flushTick:
+			b.sendBatch()
+		case <-statTick:
+			b.sendStats()
+		case <-b.stop:
+			break
+		default:
+			if len(b.records) >= b.batchSize {
```

Comment:
So yeah… after trying it again I’m reminded why this isn’t as simple as it seems ;) at least I don’t see a simple way to do it.

A few complicating factors:

* I want `Add` to not block. So therefore the `send` channel would need to be buffered so it’s non-blocking. But then that means that every time Add gets called, another value gets added to the `send` buffer, even if the buffer already contained a full batch a moment ago on prior calls to `Add`, which means that the main goroutine will receive multiple spurious signals on that channel, not just one every time the record buffer actually has reached the batch threshold. Which then means that the main goroutine, when it receives on `send`, it can’t just proceed — it has to check the buffer size again. So there’s a lot of checking going on. And none of this feels all that simple.

* We actually do want/need to do some form of polling because we need the Producer to retry if errors occur, and/or if a request partially fails and some records are returned to the buffer. Those cases don’t use `Add` to return records to the buffer, for a few reasons. And so when those cases occur, I don’t want those records to just get stuck sitting there in the buffer — as long as there’s a full batch’s worth in the buffer, I want the producer to keep trying to send them until either they succeed or they pass the `maxAttemptsPerRecord` threshold. ~~So polling seems like the way to go to cover those requirements. I suppose we *could* use yet another `time.Ticker` to do the polling, but I’m not sure that’s actually any better — it would actually use more resources and be more symbols to understand and think about, increasing the cognitive load of the code.~~ I figured out that we could do this without polling — by just sending another "send" signal whenever returning records to the buffer.

Also, a thought: if using the `default` case of a `select` was bad or non-idiomatic, I would think it wouldn’t be supported and wouldn’t be described as a normal practice in the various go team docs, talks, and posts. But it is.

Anyway, I’ve just tried using a `send` channel to trigger the send-when-a-batch-is-ready behavior and ~~another `Ticker` to do the polling,~~ and it’s just not working as I expected it. The tests are failing in different ways and sometimes hanging when I run them — there seem to be a bunch of race conditions going on maybe —  and I can’t figure out why… at this point this path of inquiry doesn’t seem worth the time and effort to me. I’ll save the changes in a stash if you want to take a look together, let me know.